tiledb/api/src/array/enumeration/mod.rs

```diff
@@ -0,0 +1,595 @@
+        // we drop self at the end of this method after returning from
+        // tiledb_enumeration_alloc.
+        self.capi_return(unsafe {
+            ffi::tiledb_enumeration_extend(
+                self.context.capi(),
+                self.capi(),
+                data.as_ptr() as *const std::ffi::c_void,
+                std::mem::size_of_val(data) as u64,
+                offsets_ptr as *const std::ffi::c_void,
+                offsets_len,
+                &mut c_new_enmr,
+            )
+        })?;
+
+        Ok(Enumeration {
+            context: self.context,
+            raw: RawEnumeration::Owned(c_new_enmr),
+        })
+    }
+}
+
+impl<'ctx> Debug for Enumeration<'ctx> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        let name = self.name().map_err(|_| fmt::Error)?;
+        let dtype = self.datatype().map_err(|_| fmt::Error)?;
+
+        let dtype_string =
+            dtype.to_string().unwrap_or("<unknown datatype>".to_owned());
+        let cell_val_num = self.cell_val_num().map_err(|_| fmt::Error)?;
+        let ordered = self.ordered().map_err(|_| fmt::Error)?;
+
+        let json = json!({
+            "name": name,
+            "datatype": dtype_string,
+            "cell_val_num": cell_val_num,
+            "ordered": ordered,
+            "values": [], // TODO: Render values
+        });
+        write!(f, "{}", json)
+    }
+}
+
+impl<'c1, 'c2> PartialEq<Enumeration<'c2>> for Enumeration<'c1> {
+    fn eq(&self, other: &Enumeration<'c2>) -> bool {
+        eq_helper!(self.name(), other.name());
+        eq_helper!(self.datatype(), other.datatype());
+        eq_helper!(self.cell_val_num(), other.cell_val_num());
+        eq_helper!(self.ordered(), other.ordered());
+        eq_helper!(self.data(), other.data());
+
+        // Can't use eq_helper here as offsets are considered equal when both
+        // are None.
+        let offsets_match = match (self.offsets(), other.offsets()) {
+            (Ok(Some(mine)), Ok(Some(theirs))) => mine == theirs,
+            (Ok(None), Ok(None)) => true,
+            _ => false,
+        };
+        if !offsets_match {
+            return false;
+        }
+
+        true
+    }
+}
+
+pub struct Builder<'ctx, 'data, 'offsets> {
+    context: &'ctx Context,
+    name: String,
+    dtype: Datatype,
+    cell_val_num: u32,
+    ordered: bool,
+    data: &'data [u8],
+    offsets: Option<&'offsets [u64]>,
+}
+
+impl<'ctx, 'data, 'offsets> ContextBound<'ctx>
+    for Builder<'ctx, 'data, 'offsets>
+{
+    fn context(&self) -> &'ctx Context {
+        self.context
+    }
+}
+
+impl<'ctx, 'data, 'offsets> Builder<'ctx, 'data, 'offsets> {
+    pub fn new<T: 'static>(
+        context: &'ctx Context,
+        name: &str,
+        dtype: Datatype,
+        data: &'data [T],
+        offsets: Option<&'offsets [u64]>,
+    ) -> Self {
+        let data = unsafe {
+            std::slice::from_raw_parts(
+                data.as_ptr() as *const std::ffi::c_void as *const u8,
```

Comment:
This is definitely wrong if `T` is `String`, but I think we can put this in the usual bucket of "things we will worry about later".

I created sc#44537 and will drop a note about this there.

Reply:
Done. That function now only accepts the `u{8,16,32,64}`, `i{8,16,32,64}`, and `f{32,46}`